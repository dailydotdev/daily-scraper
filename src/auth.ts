import {
  FastifyInstance,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  FastifyRequest,
} from 'fastify';
import fp from 'fastify-plugin';

declare module 'fastify' {
  /* eslint-disable @typescript-eslint/no-unused-vars */
  interface FastifyRequest {
    userId?: string;
    premium?: boolean;
  }
  /* eslint-enable @typescript-eslint/no-unused-vars */
}

interface Options {
  secret: string;
}

const plugin = async (
  fastify: FastifyInstance,
  opts: Options,
): Promise<void> => {
  // Machine-to-machine authentication
  fastify.addHook('preHandler', async (req) => {
    if (
      req.headers['authorization'] === `Service ${opts.secret}` &&
      req.headers['user-id'] &&
      req.headers['logged-in'] === 'true'
    ) {
      req.userId = req.headers['user-id'] as string;
      req.premium = req.headers.premium === 'true';
    } else {
      delete req.headers['user-id'];
      delete req.headers['logged-in'];
    }
  });
};

export default fp(plugin, {
  name: 'authentication',
});
