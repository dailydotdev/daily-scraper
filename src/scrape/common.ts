import * as puppeteer from 'puppeteer';

export type PageType = 'youtube' | 'generic';

export const getPageType = (page: puppeteer.Page): PageType => {
  const url = page.url();
  if (url.includes('youtube.com/channel')) {
    return 'youtube';
  }
  return 'generic';
};
