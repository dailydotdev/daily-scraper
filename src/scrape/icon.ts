import * as puppeteer from 'puppeteer';
import * as url from 'url';

interface Icon {
  url: string;
  size: number;
}

export const scrapeIcon = async (
  page: puppeteer.Page,
): Promise<string | null> => {
  const icons = await page.$$eval('link[rel="icon"]', (el) =>
    el.map(
      (x): Icon => ({
        url: x.getAttribute('href'),
        size: parseInt(x.getAttribute('sizes')?.split('x')?.[0] || '0'),
      }),
    ),
  );
  // Take the best resolution icon
  let selected = icons.sort((a, b): number => b.size - a.size)?.[0]?.url;
  if (!selected) {
    [selected] = await page.$$eval('link[rel*="apple-touch-icon"]', (el) =>
      el.map((x): string => x.getAttribute('href')),
    );
  }
  if (!selected) {
    [selected] = await page.$$eval(
      'meta[name*="msapplication-TileImage"]',
      (el) => el.map((x): string => x.getAttribute('content')),
    );
  }
  if (!selected) {
    return null;
  }
  return url.resolve(page.url(), selected);
};
