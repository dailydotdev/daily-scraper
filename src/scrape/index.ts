import * as puppeteer from 'puppeteer';

type Scraper<T> = (
  page: puppeteer.Page,
  res: puppeteer.HTTPResponse,
) => Promise<T>;

export async function scrape<T>(
  url: string,
  browser: puppeteer.Browser,
  scraper: Scraper<T>,
): Promise<T> {
  const page = await browser.newPage();
  try {
    const res = await page.goto(url, {
      waitUntil: 'load',
      timeout: 20000,
    });
    const ret = await scraper(page, res);
    await page.close();
    return ret;
  } catch (err) {
    await page.close();
    throw err;
  }
}

export * from './icon';
export * from './rss';
export * from './siteName';
