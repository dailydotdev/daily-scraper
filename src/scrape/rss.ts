import { Readable } from 'stream';
import * as puppeteer from 'puppeteer';
import * as url from 'url';
import * as FeedParser from 'feedparser';

export interface RSS {
  url: string;
  title: string;
}

export interface RSSFeed {
  meta: FeedParser.Meta;
  items: FeedParser.Item[];
}

export const scrapeRssLink = async (page: puppeteer.Page): Promise<RSS[]> => {
  let rss = await page.$$eval('link[rel="alternate"][type*="xml"]', (el) =>
    el.map(
      (x): RSS => ({
        url: x.getAttribute('href'),
        title: x.getAttribute('title'),
      }),
    ),
  );
  if (rss.length > 1) {
    rss = rss.filter((x) => !!x.title);
  } else if (rss.length > 0 && !rss[0].title) {
    rss[0].title = 'RSS';
  }
  return rss.map((x) => ({ ...x, url: url.resolve(page.url(), x.url) }));
};

export const readRssFeed = async (
  page: puppeteer.Page,
  res: puppeteer.HTTPResponse,
): Promise<RSSFeed> => {
  const contentType = res.headers()['content-type'];
  if (contentType.indexOf('xml') > -1) {
    const feed = await res.text();
    const parser = new FeedParser({ normalize: true, addmeta: true });

    const textStream = new Readable();
    textStream.push(feed, 'utf-8');
    textStream.push(null, 'utf-8');

    let rssFeed: RSSFeed;
    const promise = new Promise<RSSFeed>((resolve, reject) => {
      parser.on('readable', function () {
        rssFeed = { meta: this.meta, items: [] };
        let item: FeedParser.Item;
        while ((item = this.read())) {
          rssFeed.items.push(item);
        }
      });
      parser.on('error', reject);
      parser.on('end', () => resolve(rssFeed));
    });
    textStream.pipe(parser);
    return promise;
  }
  throw new Error('invalid content type');
};
