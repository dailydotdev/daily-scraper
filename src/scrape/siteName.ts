import * as puppeteer from 'puppeteer';

export const scrapeSiteName = async (
  page: puppeteer.Page,
): Promise<string | null> => {
  const name = await page.$$eval('meta[property="og:site_name"]', (el) =>
    el.map((x): string => x.getAttribute('content')),
  );
  return name?.[0] || null;
};
