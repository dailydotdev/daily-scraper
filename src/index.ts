import * as fastify from 'fastify';
import { FastifyInstance } from 'fastify';
import * as helmet from 'fastify-helmet';
import * as fastJson from 'fast-json-stringify';
import * as rateLimit from 'fastify-rate-limit';
import * as caching from 'fastify-caching';
import * as abstractCache from 'abstract-cache';
import * as puppeteer from 'puppeteer';

import './config';
import trace from './trace';
import auth from './auth';
import {
  scrape,
  scrapeIcon,
  scrapeRssLink,
  scrapeSiteName,
  readRssFeed,
  RSS,
} from './scrape';

export const stringifyHealthCheck = fastJson({
  type: 'object',
  properties: {
    status: {
      type: 'string',
    },
  },
});

interface ScrapeSourceWebsite {
  type: 'website';
  website?: string;
  rss: RSS[];
  logo?: string;
  name?: string;
}

interface ScrapeSourceRSS {
  type: 'rss';
  rss: string;
  website: string;
}

interface ScrapeSourceUnavailable {
  type: 'unavailable';
}

const scrapeSource = async (
  page: puppeteer.Page,
): Promise<ScrapeSourceWebsite> => {
  const [logo, rss, name] = await Promise.all([
    scrapeIcon(page),
    scrapeRssLink(page),
    scrapeSiteName(page),
  ]);
  return {
    type: 'website',
    website: page.url(),
    logo,
    rss,
    name,
  };
};

export default function app(): FastifyInstance {
  const isProd = process.env.NODE_ENV === 'production';

  const app = fastify({
    logger: true,
    disableRequestLogging: true,
    trustProxy: isProd,
  });

  app.register(helmet);
  app.register(trace, { enabled: isProd });
  app.register(auth, { secret: process.env.ACCESS_SECRET });

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  if (process.env.NODE_ENV !== 'test') {
    app.register(rateLimit, {
      max: 5,
      timeWindow: '1 minute',
      prefix: '/scrape',
    });
  }

  app.register(caching, {
    privacy: 'public',
    expiresIn: 3600,
    cache: abstractCache({ useAwait: false }),
    prefix: '/scrape',
  });

  app.get('/scrape/source', async (req, res) => {
    const { url } = req.query;
    if (!url || !url.length) {
      return res.status(400).send();
    }
    try {
      let data = await scrape(
        url,
        async (page, res): Promise<ScrapeSourceWebsite | ScrapeSourceRSS> => {
          if (res.status() >= 400) {
            throw new Error('unexpected response');
          }
          const [source, rss] = await Promise.all([
            scrapeSource(page),
            readRssFeed(page, res).catch(() => null),
          ]);
          if (rss) {
            return { type: 'rss', rss: page.url(), website: rss?.meta?.link };
          }
          return source;
        },
      );
      if (data.type === 'rss') {
        const url = data.rss;
        try {
          data = await scrape(data.website, scrapeSource);
        } catch (err) {
          data = { type: 'website', rss: [] };
        }
        data.rss = [{ title: 'RSS', url }];
      }
      res.status(200).send(data);
    } catch (err) {
      req.log.warn({ err, url }, 'failed to scrape');
      res.status(200).send({ type: 'unavailable' });
    }
  });

  return app;
}
