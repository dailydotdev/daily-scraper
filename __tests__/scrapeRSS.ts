import { scrape, scrapeRssLink } from '../src/scrape';
import { FastifyInstance } from 'fastify';
import { setupStaticServer } from './helpers';

let app: FastifyInstance;

beforeAll(async () => {
  app = await setupStaticServer();
});

afterAll(() => app.close());

it('should scrape a single rss', async () => {
  const res = await scrape(
    'http://localhost:6789/singleRSS.html',
    scrapeRssLink,
  );
  expect(res).toMatchSnapshot();
});

it('should scrape multiple rss', async () => {
  const res = await scrape(
    'http://localhost:6789/multipleRSS.html',
    scrapeRssLink,
  );
  expect(res).toMatchSnapshot();
});

it('should scrape relative rss', async () => {
  const res = await scrape(
    'http://localhost:6789/relativeRSS.html',
    scrapeRssLink,
  );
  expect(res).toMatchSnapshot();
});

it('should not find rss', async () => {
  const res = await scrape('http://localhost:6789/noMeta.html', scrapeRssLink);
  expect(res).toMatchSnapshot();
});
