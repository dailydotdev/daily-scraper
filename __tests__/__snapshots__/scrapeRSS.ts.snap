// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not find rss 1`] = `Array []`;

exports[`should scrape a single rss 1`] = `
Array [
  Object {
    "title": "RSS",
    "url": "https://daily.dev/rss.xml",
  },
]
`;

exports[`should scrape multiple rss 1`] = `
Array [
  Object {
    "title": "RSS",
    "url": "https://daily.dev/rss.xml",
  },
  Object {
    "title": "RSS2",
    "url": "https://daily.dev/rss2.xml",
  },
  Object {
    "title": "RSS3",
    "url": "https://daily.dev/rss3.xml",
  },
]
`;

exports[`should scrape relative rss 1`] = `
Array [
  Object {
    "title": "RSS",
    "url": "http://localhost:6789/rss.xml",
  },
]
`;
